README.md
# sdx-validate

[![Build Status](https://github.com/ONSdigital/sdx-validate/workflows/Build/badge.svg)](https://github.com/ONSdigital/sdx-validate) [![codecov](https://codecov.io/gh/ONSdigital/sdx-validate/branch/master/graph/badge.svg)](https://codecov.io/gh/ONSdigital/sdx-validate)

The sdx-validate app is used within the Office National of Statistics (ONS) for validating Survey Data Exchange (SDX) Surveys.

## Installation
This application presently installs required packages from requirements files:
- `requirements.txt`: packages for the application, with hashes for all packages: see https://pypi.org/project/hashin/
- `test-requirements.txt`: packages for testing and linting

It's also best to use `pyenv` and `pyenv-virtualenv`, to build in a virtual environment with the currently recommended version of Python.  To install these, see:
- https://github.com/pyenv/pyenv
- https://github.com/pyenv/pyenv-virtualenv
- (Note that the homebrew version of `pyenv` is easiest to install, but can lag behind the latest release of Python.)

### Getting started
Once your virtual environment is set, install the requirements:
```shell
$ make build
```

To test, first run `make build` as above, then run:
```shell
$ make test
```

## Usage

Start sdx-validate service using the following command:

```shell
$ python server.py
```

If you've built the image under docker, you can start using the following:

```shell
$ docker run -p 5000:5000 sdx-validate
```

sdx-validate by default binds to port 5000 on localhost and exposes two endpoints:
- '/validate' returns a json response with attributes "valid" and "error" to indicate whether the json is valid and if not what error has been thrown during validation.
- '/healthcheck' acts as a health check endpoint and returns a json response with a key/value pairs describing the service state.

### Example

The example below uses the Python library [requests](https://github.com/kennethreitz/requests) to confirm some data is valid using sdx-validate.

```python
import requests

data_to_validate = '''{
   "type": "uk.gov.ons.edc.eq:surveyresponse",
   "origin": "uk.gov.ons.edc.eq",
   "survey_id": "023",
   "version": "0.0.1",
   "collection": {
     "exercise_sid": "hfjdskf",
     "instrument_id": "0203",
     "period": "0216"
   },
   "submitted_at": "2016-03-12T10:39:40Z",
   "metadata": {
     "user_id": "789473423",
     "ru_ref": "12345678901A"
   },
   "data": {
     "11": "01/04/2016",
     "12": "31/10/2016",
     "20": "1800000",
     "51": "84",
     "52": "10",
     "53": "73",
     "54": "24",
     "50": "205",
     "22": "705000",
     "23": "900",
     "24": "74",
     "25": "50",
     "26": "100",
     "21": "60000",
     "27": "7400",
     "146": "some comment"
   }
}'''

r = requests.post('http://127.0.0.1:5000/validate', data=data_to_validate)

is_valid = r.data.valid
```

## License

Copyright (c) 2016 Crown Copyright (Office for National Statistics)

Released under MIT license, see [LICENSE](LICENSE) for details.


CHANGELOG.md
### Unreleased

### 2.20.0 2021-02-22
 - 007.0009 and 007.0010 survey added to validation

### 2.19.0 2020-10-20
 - Add MES survey formtype 0001 (092)
 
### 2.18.0 2020-09-07
 - Add EPE survey formtypes 0003 and 0004 (147)
 - Add openapi.yml

### 2.17.0 2020-06-09
 - Add construction survey formtype 0002 (228)

### 2.16.2 2020-05-21
 - Updated packages

### 2.16.1 2020-05-13
 - Remove Cloudfoundry deployment files

### 2.16.0 2020-04-17
 - Add covid-19 survey formtype 0002 (283)
 - Add construction survey (228)

### 2.15.0 2020-03-20
 - Add covid-19 survey formtype (283)

### 2.14.2 2020-03-18
 - Remove changes made in 2.13.0.  It didn't work as expected.

### 2.14.1 2020-03-03
 - Amend vacancies formtypes to correct ones

### 2.14.0 2020-01-30
 - Add Vacancies formtypes (182, 183, 184, 185)

### 2.13.0 2020-01-27
 - Validate submission for null characters  (`\u0000`) and fail validation if present

### 2.12.0 2019-11-25
 - Add E-commerce formtypes 0001 and 0002

### 2.11.2 2019-11-08
 - Add python 3.8 to travis builds

### 2.11.1 2019-09-26
 - Update packages, including Werkzeug that had a security vulnerability

### 2.11.0 2019-05-31
 - Added support for QSI (Stocks)
 - Remove python 3.4 and 3.5 from travis builds
 - Add python 3.7 to travis builds

### 2.10.0 2019-02-19
 - Add support for QPSES

### 2.9.0 2019-02-08
 - Added support for UKIS
 - Add support for NI version of UKIS

### 2.8.0 2018-12-12
 - Added support for E-Commerce

### 2.7.0 2018-11-13
 - Added support for D-Trades

### 2.6.0 2018-10-22
 - Added support for LMS 2.
 - Made case_ref optional for version 0.0.2 surveys

### 2.5.0 2018-09-11
 - Add support for QCAS

### 2.4.0 2018-08-15
 - Update dependency versions
 - Add support for 'started_at' field in payload

### 2.3.0 2018-06-20
 - Support LMS survey

### 2.2.0 2018-05-21
 - Support MBS surveys

### 2.1.0 2018-03-06
 - Support case_id and case_ref fields

### 2.0.0 2017-01-04
  - Remove support for heartbeat survey
  - Add /info healthcheck endpoint

### 1.12.0 2017-11-21
  - Add all service specific config to config file
  - Change to use pytest to improve test output and also changes to improve code coverage stats
  - Add Cloudfoundry deployment files
  - Remove sdx-common logging and bring it back into the repo

### 1.11.0 2017-09-25
  - Permit reference period defaults in metadata

### 1.10.1 2017-09-11
  - Update version number in app

### 1.10.0 2017-09-11
  - Processing feedback surveys

### 1.9.1
  - Ensure integrity and version of library dependencies
  - Accept floats and ints in data key

### 1.8.0 2017-07-25
  - Change all instances of ADD to COPY in Dockerfile
  - Remove use of SDX_HOME variable in makefile
  - Missing or blank survey ids now return a 400 rather than 500
  - Add support for feedback survey responses

### 1.7.0 2017-07-10
  - Update and pin version of sdx-common to 0.7.0

### 1.6.0 2017-06-19
  - Add license
  - Add codacy badge
  - Adding sdx-common functionality
  - Updating logger format using sdx-common
  - Prepare for MCI refresh
  - Add support for codecov to see unit test coverage

### 1.5.0 2017-05-04
  - Support 'flushed' flag in schema definitions. Maintain backwards compatibility with older schema versions.
  - Validate ids based on schema
  - Validate data for different survey types
  - Support 'completed' flag in schema definitions. See the [schema definitions](https://github.com/ONSdigital/ons-schema-definitions/blob/master/docs/electronic_questionnaire_to_data_exchange.rst) for more information.

## 1.4.1 2017-03-15
  - Log version number on startup

### 1.4.0 2017-02-16
  - Add change log
  - Update known instruments to correlate to specific surveys
  - Add QBS details for validation
  - Add UKIS details for validation
  - Add MWSS details for validation

### 1.3.0 2016-12-13
  - Add support for validating version 0.0.2 schemas

### 1.2.1 2016-11-28
  - Remove sensitive data from logging

### 1.2.0 2016-10-19
  - Add new `/healthcheck` endpoint

### 1.1.0 2016-09-19
  - Add configurable log level

### 1.0.1 2016-08-19
  - Fix issue with paradata element

### 1.0.0 2016-08-05
  - Initial release


ci/test.sh
#!/bin/bash

set -e -x

pushd sdx-validate
  make
  make test
popd

main_test.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	. "github.com/smartystreets/goconvey/convey"
)

var (
	goodData = `{
       "tx_id": "0f534ffc-9442-414c-b39f-a756b4adc6cb",
       "type": "uk.gov.ons.edc.eq:surveyresponse",
       "origin": "uk.gov.ons.edc.eq",
       "survey_id": "023",
       "version": "0.0.1",
       "collection": {
         "exercise_sid": "hfjdskf",
         "instrument_id": "0203",
         "period": "0216"
       },
       "submitted_at": "2016-03-12T10:39:40Z",
       "metadata": {
         "user_id": "789473423",
         "ru_ref": "12345678901A"
       },
       "data": {
         "11": "01/04/2016",
         "12": "31/10/2016",
         "20": "1800000",
         "51": "84",
         "52": "10",
         "53": "73",
         "54": "24",
         "50": "205",
         "22": "705000",
         "23": "900",
         "24": "74",
         "25": "50",
         "26": "100",
         "21": "60000",
         "27": "7400",
         "146": "some comment"
       },
       "paradata": {}
    }`
)

func sendToEndpoint(data string) (code int, valid bool, err error) {
	log.Println("[test] POSTING to endpoint")
	req, err := http.NewRequest("POST", "/validate", bytes.NewBufferString(data))
	if err != nil {
		log.Fatal("Unable to create new HTTP request")
		return
	}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(validateHandler)

	handler.ServeHTTP(rr, req)

	body := rr.Body.Bytes()
	if body == nil {
		err = errors.New("Empty body returned from endpoint")
		log.Printf("%v", err)
		return
	}

	r := make(map[string]bool)
	err = json.Unmarshal(body, &r)
	if err != nil {
		log.Fatalf("Failed to unmarshal reply: %v", err)
		return
	}

	code = rr.Code
	valid = r["valid"]
	return
}

func amendSurveyValue(data, key string, value interface{}) (string, error) {

	// var sm map[string]interface{}
	var surveyMap, level map[string]interface{}
	So(json.Unmarshal([]byte(data), &surveyMap), ShouldBeNil)

	level = surveyMap

	// Deal with the key being nested
	for i, k := range strings.Split(key, ".") {
		switch {
		case i == strings.Count(key, "."):
			switch value {
			case nil:
				delete(level, k) // if requested 'nil' then delete the key
			default:
				level[k] = value
			}
		default:
			level = level[k].(map[string]interface{})
		}
	}

	amended, err := json.Marshal(surveyMap)
	if err != nil {
		return "", err
	}

	return string(amended), nil
}

func TestFailSends500(t *testing.T) {
	Convey("Should return 500 when sent non-JSON data", t, func() {
		code, valid, err := sendToEndpoint(`rubbish`)
		if err != nil {
			t.Fatalf("Test to endpoint failed: %v [%v %v]", err, code, valid)
		}
		So(code, ShouldEqual, http.StatusInternalServerError)
		So(valid, ShouldBeFalse)
	})
}

func TestOKWithValidJSON(t *testing.T) {
	Convey("Should return valid when data is valid", t, func() {
		code, valid, err := sendToEndpoint(goodData)
		if err != nil {
			t.Fatalf("Sending to endpoint failed: %v", err)
		}
		So(code, ShouldEqual, http.StatusOK)
		So(valid, ShouldBeTrue)
	})
}

func TestFailWhenDataEmpty(t *testing.T) {
	Convey("When data is empty", t, func() {
		code, valid, err := sendToEndpoint("")
		if err != nil {
			t.Fatalf("Sending to endpoint failed: %v", err)
		}
		So(code, ShouldEqual, http.StatusInternalServerError) // TODO should this be 500? Or should it be 200 OK?
		So(valid, ShouldBeFalse)
	})
}

func TestTransactionIDIsOptional(t *testing.T) {
	Convey("When sending data without transaction id (tx_id)", t, func() {

		sa, err := amendSurveyValue(goodData, "tx_id", nil)
		if err != nil {
			t.Fatalf("Error attempting to delete tx_id from survey data")
		}

		code, valid, err := sendToEndpoint(sa)
		if err != nil {
			t.Fatalf("Sending to endpoint failed: %v", err)
		}
		So(code, ShouldEqual, http.StatusOK)
		So(valid, ShouldBeTrue)
	})
}

func TestInvalidItems(t *testing.T) {

	type invalidItemTestSet struct {
		Name   string
		Field  string
		Values []string
	}

	ts := []invalidItemTestSet{
		invalidItemTestSet{
			Name:   "version",
			Field:  "version",
			Values: []string{"0.0.0", "9.9.9"},
		},
		invalidItemTestSet{
			Name:   "survey ids",
			Field:  "survey_id",
			Values: []string{"999", "029"},
		},
		invalidItemTestSet{
			Name:   "instrument ids",
			Field:  "collection.instrument_id",
			Values: []string{"999", "111"},
		},
	}

	Convey("When using invalid values", t, func() {
		for _, set := range ts {
			Convey(fmt.Sprintf("For %v", set.Name), func() {

				for _, value := range set.Values {
					sl, err := amendSurveyValue(goodData, set.Field, value)

					if err != nil {
						t.Fatalf("Error attempting to amend survey value: %v", err)
					}

					code, valid, err := sendToEndpoint(sl)
					if err != nil {
						t.Fatalf("Error returned from endpoint: %v", err)
					}

					Convey(fmt.Sprintf("With %s set to %s", set.Field, value), func() {
						So(code, ShouldEqual, http.StatusOK)
						So(valid, ShouldBeFalse)
					})
				}
			})
		}
	})
}

func TestNonGUIDTransactionIDIsInvalid(t *testing.T) {
	Convey("When submitting data", t, func() {
		for _, v := range []string{
			"999", // Bad
			"f81d4fae-7dec-11d0-a765-00a0c91e6bf",  // Last char missing
			"f81d4fae-7dec-11d0-z765-00a0c91e6bf6", // Wrong char in 17th position}
		} {
			Convey(fmt.Sprintf("With tx_id of '%v'", v), func() {
				sa, err := amendSurveyValue(goodData, "tx_id", v)
				if err != nil {
					t.Fatalf("Error attempting to set tx_id to %v", v)
				}
				code, valid, err := sendToEndpoint(sa)
				if err != nil {
					t.Fatalf("Error returned from endpoint: %v", err)
				}
				So(code, ShouldEqual, http.StatusOK)
				So(valid, ShouldBeFalse)
			})
		}
	})
}


tests/test_service.py
from server import app
import server

import unittest
import json


class TestValidateService(unittest.TestCase):

    validate_endpoint = "/validate"
    healthcheck_endpoint = "/healthcheck"
    message = {
        '0.0.1': '''{
           "tx_id": "0f534ffc-9442-414c-b39f-a756b4adc6cb",
           "type": "uk.gov.ons.edc.eq:surveyresponse",
           "origin": "uk.gov.ons.edc.eq",
           "survey_id": "023",
           "completed": true,
           "flushed": false,
           "version": "0.0.1",
           "collection": {
             "exercise_sid": "hfjdskf",
             "instrument_id": "0203",
             "period": "0216"
           },
           "submitted_at": "2016-03-12T10:39:40Z",
           "metadata": {
             "user_id": "789473423",
             "ru_ref": "12345678901A"
           },
           "data": {
             "11": "01/04/2016",
             "12": "31/10/2016",
             "20": "1800000",
             "51": 84.00,
             "52": 10,
             "53": "73",
             "54": "24",
             "50": "205",
             "22": "705000",
             "23": "900",
             "24": "74",
             "25": "50",
             "26": "100",
             "21": "60000",
             "27": "7400",
             "146": "some comment"
           },
           "paradata": {}
        }''',

        '0.0.2': '''{
           "tx_id": "0f534ffc-9442-414c-b39f-a756b4adc6cb",
           "case_ref": "1000000000000001",
           "case_id": "4c0bc9ec-06d4-4f66-88b6-2e42b79f17b3",
           "type": "uk.gov.ons.edc.eq:surveyresponse",
           "origin": "uk.gov.ons.edc.eq",
           "survey_id": "census",
           "completed": false,
           "flushed": true,
           "version": "0.0.2",
           "collection": {
             "exercise_sid": "hfjdskf",
             "instrument_id": "household",
             "period": "0216"
           },
           "submitted_at": "2016-03-12T10:39:40Z",
           "metadata": {
             "user_id": "789473423",
             "ru_ref": "12345"
           },
           "data": [{
             "value": "Joe Bloggs",
             "block_id": "household-composition",
             "answer_id": "household-full-name",
             "group_id": "multiple-questions-group",
             "group_instance": 0,
             "answer_instance": 0
            }]
        }''',

        'feedback': '''{
               "type" : "uk.gov.ons.edc.eq:feedback",
               "origin" : "uk.gov.ons.edc.eq",
               "metadata": {
                 "user_id": "789473423",
                 "ru_ref": "432423423423"
               },
               "data": {
                 "url": "https://eq.onsdigital.uk/feedback",
                 "name": "John Appleseed",
                 "email": "john.appleseed@ons.gov.uk",
                 "message": "Feedback message string"
               },
               "submitted_at": "2016-03-07T15:28:05Z",
               "collection": {
                 "instrument_id": "0203",
                 "exercise_sid": "739",
                 "period": "2016-02-01"
               },
               "survey_id": "023",
               "tx_id": "0f534ffc-9442-414c-b39f-a756b4adc6cb",
               "version" : "0.0.1"
            }''',
    }

    def setUp(self):
        # creates a test client
        self.app = app.test_client()
        # propagate the exceptions to the test client
        self.app.testing = True

    def test_unexpected_payload_data(self):
        survey = json.loads(self.message['0.0.1'])
        survey['unexpected_data'] = 'unexpected'
        self.assertInvalid(survey)

    def validate_response(self, data):
        dumped_json = json.dumps(data)

        r = self.app.post(self.validate_endpoint, data=dumped_json)
        actual_response = json.loads(r.data.decode('UTF8'))

        return actual_response

    def assertInvalid(self, data):
        actual_response = self.validate_response(data)

        self.assertEqual(actual_response['valid'], False)
        self.assertNotIn('contains_null_character', actual_response)
        self.assertEqual(actual_response['status'], 400)

    def assertValid(self, data):
        actual_response = self.validate_response(data)
        self.assertEqual(actual_response['valid'], True)

    def test_validate_fail_sends_500(self):
        r = self.app.post(self.validate_endpoint, data='rubbish')

        self.assertEqual(r.status_code, 500)

    def test_validates_json(self):
        expected_response = json.dumps({"valid": True})
        for v in ['0.0.1', '0.0.2', 'feedback']:
            m = self.message[v]
            r = self.app.post(self.validate_endpoint, data=m)
            actual_response = json.dumps(json.loads(r.data.decode('UTF8')))
            self.assertEqual(actual_response, expected_response)

    def test_mwss_valid(self):
        survey = json.loads(self.message['0.0.1'])
        survey['survey_id'] = "134"
        survey['collection']['instrument_id'] = "0005"
        self.assertValid(survey)

    def test_mbs_valid(self):
        survey = json.loads(self.message['0.0.1'])
        survey['survey_id'] = "009"
        survey['collection']['instrument_id'] = "0255"
        self.assertValid(survey)

    def test_epe_valid(self):
        survey = json.loads(self.message['0.0.1'])
        survey['survey_id'] = "147"

        survey['collection']['instrument_id'] = "0003"
        self.assertValid(survey)

        survey['collection']['instrument_id'] = "0004"
        self.assertValid(survey)

    def test_qcas_valid(self):
        survey = json.loads(self.message['0.0.1'])
        survey['survey_id'] = "019"

        survey['collection']['instrument_id'] = "0018"
        self.assertValid(survey)

        survey['collection']['instrument_id'] = "0019"
        self.assertValid(survey)

        survey['collection']['instrument_id'] = "0020"
        self.assertValid(survey)

        survey['collection']['instrument_id'] = "0021"
        self.assertInvalid(survey)

    def test_rsi_metadata(self):
        for inst_id in ["0102", "0112", "0203", "0205", "0213", "0215"]:
            with self.subTest(inst_id=inst_id):
                survey = json.loads(self.message["0.0.1"])
                survey["survey_id"] = "023"
                survey["collection"]["instrument_id"] = inst_id
                survey["metadata"]["ref_period_start_date"] = "2016-04-01"
                survey["metadata"]["ref_period_end_date"] = "2016-10-31"
                self.assertValid(survey)

    def test_invalid_version(self):
        survey = json.loads(self.message['0.0.1'])
        with self.assertRaises(KeyError):
            server.ValidSurveyId(survey['survey_id'], '0.0.3')

    def test_missing_version(self):
        survey = json.loads(self.message['0.0.1'])
        with self.assertRaises(AttributeError):
            server.ValidSurveyId(survey['survey_id'])

    def test_valid_version(self):
        survey = json.loads(self.message['0.0.1'])
        server.ValidSurveyId(survey['survey_id'], survey['version'])

    def test_unknown_version_invalid(self):
        unknown_version = json.loads(self.message['0.0.1'])
        unknown_version['version'] = "0.0.3"

        self.assertInvalid(unknown_version)

    def test_unknown_survey_invalid(self):
        unknown_survey = json.loads(self.message['0.0.1'])
        unknown_survey['survey_id'] = "025"

        self.assertInvalid(unknown_survey)

    def test_blank_survey_invalid(self):
        unknown_survey = json.loads(self.message['0.0.1'])
        unknown_survey['survey_id'] = ""

        self.assertInvalid(unknown_survey)

    def test_missing_survey_invalid(self):
        unknown_survey = json.loads(self.message['0.0.1'])
        del unknown_survey['survey_id']

        self.assertInvalid(unknown_survey)

    def test_unknown_census_survey_invalid(self):
        unknown_survey = json.loads(self.message['0.0.2'])
        unknown_survey['survey_id'] = "025"

        self.assertInvalid(unknown_survey)

    def test_unknown_instrument_invalid(self):
        unknown_instrument = json.loads(self.message['0.0.1'])
        unknown_instrument['collection']['instrument_id'] = "999"

        self.assertInvalid(unknown_instrument)

    def test_unknown_census_instrument_invalid(self):
        unknown_instrument = json.loads(self.message['0.0.2'])
        unknown_instrument['collection']['instrument_id'] = "999"

        self.assertInvalid(unknown_instrument)

    def test_known_instrument_wrong_survey_invalid(self):
        # RSI survey_id with Census instrument_id
        known_instrument = json.loads(self.message['0.0.1'])
        known_instrument['collection']['instrument_id'] = "household"

        self.assertInvalid(known_instrument)

    def test_known_census_instrument_wrong_survey_invalid(self):
        # RSI survey_id with Census instrument_id
        known_instrument = json.loads(self.message['0.0.2'])
        known_instrument['collection']['instrument_id'] = "0203"

        self.assertInvalid(known_instrument)

    def test_known_instrument_correct_survey_valid(self):
        # RSI survey_id with RSI instrument_id
        known_instrument = json.loads(self.message['0.0.1'])
        known_instrument['collection']['instrument_id'] = "0213"

        self.assertValid(known_instrument)

    def test_known_census_instrument_correct_survey_valid(self):
        # RSI survey_id with RSI instrument_id
        known_instrument = json.loads(self.message['0.0.2'])
        known_instrument['collection']['instrument_id'] = "household"

        self.assertValid(known_instrument)

    def test_empty_data_invalid(self):
        empty_data = json.loads(self.message['0.0.1'])
        empty_data['data'] = ""

        self.assertInvalid(empty_data)

    def test_census_empty_data_invalid(self):
        empty_data = json.loads(self.message['0.0.2'])
        empty_data['data'] = ""

        self.assertInvalid(empty_data)

    def test_feedback_empty_data_invalid(self):
        empty_data = json.loads(self.message['feedback'])
        empty_data['data'] = ""

        self.assertInvalid(empty_data)

    def test_census_dict_data_invalid(self):
        dict_data = json.loads(self.message['0.0.2'])
        dict_data['data'] = {'key1': 'value1', 'key2': 'value2'}

        self.assertInvalid(dict_data)

    def test_census_list_dict_plain_data_valid(self):
        data = json.loads(self.message['0.0.2'])
        data['data'] = ["a", "b", "c", {"Some": "Thing"}]

        self.assertValid(data)

    def test_string_data_invalid(self):
        data = "abcd"

        self.assertInvalid(data)

    def test_no_data(self):
        data = None

        self.assertInvalid(data)

    def test_server_error(self):
        r = self.app.post(self.validate_endpoint, data=None)
        actual_response = json.loads(r.data.decode('UTF8'))
        self.assertEqual(actual_response['valid'], False)
        self.assertEqual(actual_response['status'], 500)

    def test_healthcheck(self):
        r = self.app.get(self.healthcheck_endpoint)
        actual_response = json.loads(r.data.decode('UTF8'))
        self.assertEqual(actual_response, {"status": "OK"})

    def test_census_binary_data_error(self):
        data = json.loads(self.message['0.0.2'])
        data['data'] = b'I am text.'

        with self.assertRaises(TypeError):
            self.validate_response(data)

    def test_non_guid_tx_id_invalid(self):
        wrong_tx = json.loads(self.message['0.0.1'])
        wrong_tx['tx_id'] = "999"

        self.assertInvalid(wrong_tx)

        # Last character missing
        wrong_tx['tx_id'] = "f81d4fae-7dec-11d0-a765-00a0c91e6bf"

        self.assertInvalid(wrong_tx)

        # Wrong character at 17th position
        wrong_tx['tx_id'] = "f81d4fae-7dec-11d0-z765-00a0c91e6bf6"

        self.assertInvalid(wrong_tx)

    def test_tx_id_optional(self):
        message = json.loads(self.message['0.0.1'])
        del message['tx_id']

        self.assertValid(message)

    def test_completed_not_boolean_fails(self):
        message = json.loads(self.message['0.0.1'])
        message['completed'] = ''

        self.assertInvalid(message)

    def test_flushed_not_boolean_fails(self):
        message = json.loads(self.message['0.0.1'])
        message['flushed'] = ''

        self.assertInvalid(message)

    def test_completed_key_missing_fails(self):
        message = json.loads(self.message['0.0.1'])
        message.pop('completed')

        self.assertRaises(KeyError, message.__getitem__, 'completed')
        self.assertValid(message)

    def test_flushed_key_missing_fails(self):
        message = json.loads(self.message['0.0.1'])
        message.pop('flushed')

        self.assertRaises(KeyError, message.__getitem__, 'flushed')
        self.assertValid(message)

    def test_case_id_and_case_ref_passes(self):
        message = json.loads(self.message['0.0.1'])
        self.assertValid(message)

    def test_case_id_not_str_fails(self):
        message = json.loads(self.message['0.0.1'])
        message['case_id'] = {}

        self.assertInvalid(message)

    def test_case_ref_not_str_fails(self):
        message = json.loads(self.message['0.0.1'])
        message['case_ref'] = {}

        self.assertInvalid(message)

    def test_valid_lms(self):
        survey = json.loads(self.message['0.0.2'])
        survey['survey_id'] = "lms"
        survey['collection']['instrument_id'] = "1"

        self.assertValid(survey)


server.py
import logging
import os

from functools import partial
from voluptuous import Schema, Required, Length, All, MultipleInvalid, Optional
from dateutil import parser
from flask import Flask, request, jsonify
from structlog import wrap_logger
from uuid import UUID

import settings


__version__ = "2.20.0"

app = Flask(__name__)

logging.basicConfig(
    format=settings.LOGGING_FORMAT,
    datefmt="%Y-%m-%dT%H:%M:%S",
    level=settings.LOGGING_LEVEL,
)

logger = wrap_logger(logging.getLogger(__name__))
logger.debug("START", version=__version__)

KNOWN_SURVEYS = {
    "0.0.1": {
        "007": [
            "0010", "0009"
        ],
        "009": [
            "0106", "0111", "0117", "0123", "0158", "0161", "0167", "0173", "0201", "0202",
            "0203", "0204", "0205", "0216", "0251", "0253", "0255", "0817", "0823", "0867",
            "0873",
        ],
        "017": [
            "0001", "0002", "0003", "0004", "0005", "0006", "0007", "0008", "0009", "0010",
            "0011", "0012", "0013", "0014", "0033", "0034", "0051", "0052", "0057", "0058",
            "0061", "0070"
        ],
        "019": ["0018", "0019", "0020"],
        "023": ["0203", "0213", "0205", "0215", "0102", "0112"],
        "092": ["0001"],
        "134": ["0005"],
        "139": ["0001"],
        "144": ["0001", "0002"],
        "147": ["0003", "0004"],
        "160": ["0002"],
        "165": ["0002"],
        "169": ["0003"],
        "182": ["0006"],
        "183": ["0006"],
        "184": ["0006"],
        "185": ["0005"],
        "187": ["0001", "0002", "0051"],
        "228": ["0001", "0002"],
        "283": ["0001", "0002"],
    },
    "0.0.2": {
        "census": ["household", "individual", "communal"],
        "lms": ["1", "2"],
        "281": ["0001"],
    },
}


# Parses a timestamp, throwing a value error
# if unrecognised


def Timestamp(value):
    return parser.parse(value)


def ValidSurveyId(value, version=None):
    if not version:
        raise AttributeError("No version number")

    if value not in KNOWN_SURVEYS[version]:
        raise ValueError("Invalid survey id")


# Parses a UUID, throwing a value error
# if unrecognised
def ValidSurveyTxId(value):
    return UUID(value, version=4)


def ValidSurveyData(data):
    if isinstance(data, dict):
        for k, v in data.items():
            if not isinstance(k, str) or not isinstance(v, (str, list, int, float)):
                raise ValueError("Invalid survey data")

    else:
        raise ValueError("Invalid survey data")


def ValidateListSurveyData(data):
    if not isinstance(data, list):
        raise ValueError("Invalid surey data")


@app.errorhandler(400)
def errorhandler_400(e):
    return client_error(repr(e))


def client_error(error=None):
    logger.error("Validation error", exception=error)
    message = {"valid": False, "status": 400, "message": error, "uri": request.url}
    resp = jsonify(message)
    resp.status_code = 400

    return resp


@app.errorhandler(500)
def server_error(e):
    logger.error("Server Error", exception=repr(e))
    message = {
        "valid": False, "status": 500, "message": "Internal server error: " + repr(e)
    }
    resp = jsonify(message)
    resp.status_code = 500

    return resp


@app.route("/validate", methods=["POST"])
def validate():
    try:
        json_data = request.get_json(force=True)

        response_type = str(json_data["type"])

        if response_type.find("feedback") == -1:
            version = json_data["version"]

            schema = get_schema(version)

            if schema is None:
                return client_error("Unsupported schema version '%s'" % version)

            metadata = json_data.get("metadata")
            bound_logger = logger.bind(survey_id=json_data.get("survey_id"),
                                       tx_id=json_data.get("tx_id"),
                                       user_id=metadata.get("user_id"),
                                       ru_ref=metadata.get("ru_ref"))

            bound_logger.debug("Validating json against schema")
            schema(json_data)

            survey_id = json_data.get("survey_id")
            if survey_id not in KNOWN_SURVEYS.get(version, {}):
                bound_logger.debug("Survey id is not known", survey_id=survey_id)
                return client_error(f"Unsupported survey '{survey_id}'")

            instrument_id = json_data.get("collection", {}).get("instrument_id")
            if instrument_id not in KNOWN_SURVEYS.get(version, {}).get(survey_id, []):
                bound_logger.debug("Instrument ID is not known", survey_id=survey_id)
                return client_error(f"Unsupported instrument '{instrument_id}'")

        else:
            schema = get_schema("feedback")

            bound_logger = logger.bind(response_type="feedback", tx_id=json_data.get("tx_id"))
            bound_logger.debug("Validating json against schema")
            schema(json_data)

        bound_logger.debug("Success")

    except (MultipleInvalid, KeyError, TypeError, ValueError) as e:
        logger.error("Client error", error=e)
        return client_error(str(e))

    except Exception as e:
        logger.error("Server error", error=e)
        return server_error(e)

    return jsonify({"valid": True})


@app.route("/info", methods=["GET"])
@app.route("/healthcheck", methods=["GET"])
def healthcheck():
    return jsonify({"status": "OK"})


def get_schema(version):

    if version == "0.0.1":
        valid_survey_id = partial(ValidSurveyId, version="0.0.1")

        collection_s = Schema(
            {
                Required("period"): str,
                Required("exercise_sid"): str,
                Required("instrument_id"): All(str, Length(max=4)),
            }
        )

        metadata_s = Schema(
            {
                Required("user_id"): str,
                Required("ru_ref"): All(str, Length(12)),
                Optional("ref_period_start_date"): Timestamp,
                Optional("ref_period_end_date"): Timestamp,
            }
        )

        schema = Schema(
            {
                Required("type"): "uk.gov.ons.edc.eq:surveyresponse",
                Required("version"): "0.0.1",
                Optional("tx_id"): All(str, ValidSurveyTxId),
                Required("origin"): "uk.gov.ons.edc.eq",
                Required("survey_id"): All(str, valid_survey_id),
                Optional("case_id"): str,
                Optional("case_ref"): str,
                Optional("completed"): bool,
                Optional("flushed"): bool,
                Optional("started_at"): Timestamp,
                Required("submitted_at"): Timestamp,
                Required("collection"): collection_s,
                Required("metadata"): metadata_s,
                Required("data"): ValidSurveyData,
                Optional("paradata"): object,
            }
        )
        return schema

    elif version == "0.0.2":
        valid_survey_id = partial(ValidSurveyId, version="0.0.2")

        collection_s = Schema(
            {
                Required("period"): str,
                Required("exercise_sid"): str,
                Required("instrument_id"): All(str, Length(max=10)),
            }
        )

        metadata_s = Schema(
            {
                Required("user_id"): str,
                Optional("ru_ref"): str,
                Optional("ref_period_start_date"): Timestamp,
                Optional("ref_period_end_date"): Timestamp,
            }
        )

        schema = Schema(
            {
                Required("type"): "uk.gov.ons.edc.eq:surveyresponse",
                Required("version"): "0.0.2",
                Optional("tx_id"): All(str, ValidSurveyTxId),
                Required("origin"): "uk.gov.ons.edc.eq",
                Required("survey_id"): All(str, valid_survey_id),
                Optional("completed"): bool,
                Optional("flushed"): bool,
                Optional("started_at"): Timestamp,
                Required("submitted_at"): Timestamp,
                Required("case_id"): str,
                Optional("case_ref"): str,
                Required("collection"): collection_s,
                Required("metadata"): metadata_s,
                Required("data"): ValidateListSurveyData,
                Optional("paradata"): object,
            }
        )
        return schema

    elif version == "feedback":

        collection_s = Schema(
            {
                Required("period"): str,
                Required("exercise_sid"): str,
                Required("instrument_id"): All(str, Length(max=10)),
            }
        )

        schema = Schema(
            {
                Required("type"): "uk.gov.ons.edc.eq:feedback",
                Optional("tx_id"): All(str, ValidSurveyTxId),
                Required("origin"): "uk.gov.ons.edc.eq",
                Required("survey_id"): str,
                Required("version"): str,
                Optional("completed"): bool,
                Optional("flushed"): bool,
                Optional("started_at"): Timestamp,
                Required("submitted_at"): Timestamp,
                Required("collection"): collection_s,
                Required("data"): ValidSurveyData,
                Optional("metadata"): object,
                Optional("paradata"): object,
            }
        )
        return schema

    else:
        return None


if __name__ == "__main__":
    port = int(os.getenv("PORT"))
    app.run(debug=True, host="0.0.0.0", port=port)
